File: pset2/problems/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

// Prototypes
int *count_text(string text);
void get_index(int letters, int words, int sentences);

int main(void) {
    string text = get_string("Text: ");
    int *count_result = count_text(text);

    int letters = count_result[0];
    int words = count_result[1];
    int sentences = count_result[2];
    int len = count_result[3];
    printf("Letters %i\n", letters);
    printf("Words %i\n", words);
    printf("Sentences %i\n", sentences);

    get_index(letters, words, sentences);
}

int *count_text(string text) {
    static int result[4];
    int i = 0;
    int count_letters = 0;
    int count_words = 0;
    int count_sentences = 0;
    int is_word = 0;
    int beginning = 0;

    while (text[i] != '\0') {
        // Count letters, words and handle the beginning of a word
        if (isalpha(text[i])) {
            count_letters++;
            i++;
            if (!is_word) {
                count_words++;
                is_word = 1;
            }
        }
        // Handle spaces to mark the end of a word
        else if (isspace(text[i])) {
            i++;
            is_word = 0;
        }
        // Count sentences and mark the end of a word
        else if (text[i] == '?' || text[i] == '.' || text[i] == '!') {
            count_sentences++;
            is_word = 0;
            i++;
        }
        // Skip non-alphabetic and non-space characters
        else {
            i++;
        }
    }
    result[0] = count_letters;
    result[1] = count_words;
    result[2] = count_sentences;
    result[3] = i;

    return result;
}

void get_index(int letters, int words, int sentences) {
    float L = (float)letters / words * 100;
    float S = (float)sentences / words * 100;
    float index = 0.0588 * L - 0.296 * S - 15.8;

    int rounded = round(index);
    if (index < 1) {
        printf("Before Grade 1\n");
    } else if (index > 16) {
        printf("Grade 16+\n");
    } else {
        printf("Grade %i\n", rounded);
    }
}

File: pset6/problems/sentimental/readability.py
# TODO


def main():
    """Calculate and print the Coleman-Liau grade for a given text.

    This function prompts the user to input a text, analyzes it for the number
    of letters, words, and sentences, calculates the Coleman-Liau Index, and
    then prints the corresponding readability grade. The user experiences an
    interactive process to assess the readability of the provided text.

    """
    text = input("Text: ")
    letters, words, sentences = count_text(text)
    index = calculate_index(letters, words, sentences)
    print(
        f"letters: {letters}, words: {words}, sentences: {sentences}, index: {index}"
    )

    print_grade(index)


def count_text(text: str) -> tuple[int, int, int]:
    """Count the number of letters, words, and sentences in the given text.

    Parameters
    ----------
    text : str
        The input text for counting.

    Returns
    -------
    tuple[int, int, int]
        A tuple containing counts of letters, words, and sentences.
    """
    i = 0
    letters = 0
    words = 0
    sentences = 0
    is_word = False

    while i < len(text):
        # Count letters, words and handle the beginning of a word
        if text[i].isalpha():
            letters += 1
            i += 1
            if not is_word:
                words += 1
                is_word = True
        # Handle spaces to mark the end of a word
        elif text[i].isspace():
            i += 1
            is_word = False
        # Count sentences and mark the end of a word
        elif text[i] == "?" or text[i] == "." or text[i] == "!":
            sentences += 1
            is_word = False
            i += 1
        # Skip non-alphabetic and non-space characters
        else:
            i += 1

    return (letters, words, sentences)





if __name__ == "__main__":
    main()
